### PR TITLE
2891 making phTracedEnclosure thread safe

### LINKED ISSUE #2891: PhTracedEnclosure.java:36-38: Make the class thread safe....
The puzzle `2836-88180ad7` from #2836 has to be resolved: 

https://github.com/objectionary/eo/blob/7b604a832154a5ebd0e3f030bd9e6451a71afc53/eo-runtime/src/main/java/org/eolang/PhTracedEnclosure.java#L36-L38

The puzzle was created by Yegor Bugayenko on 22-Feb-24. 

Estimate: 60 minutes,  role: DEV. 

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by levBagryansky**
Closes #2891

---

## PR-Codex overview
This PR enhances thread safety in `CachedPhi` and `PhTracedEnclosure`. It adds testing for thread safety, updates documentation, and uses `ThreadLocal` for better synchronization.

### Detailed summary
- Added testing for thread safety in `CachedPhi`
- Updated documentation in `CachedPhi`
- Improved thread safety using `ThreadLocal` in `PhTracedEnclosure`
- Refactored `PhTracedEnclosure` for better synchronization

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by levBagryansky** [comment]
@maxonfjvipon please check

File: eo-runtime/src/main/java/org/eolang/PhTracedEnclosure.java [hunk 1]
```diff
@@ -29,11 +29,7 @@
 
 /**
  * Class to trace if the cage got into recursion during the dataization.
- * NOT thread-safe.
```
[the remark this answers is not in the document]
Comment (levBagryansky) [reply eo-runtime/src/main/java/org/eolang/PhTracedEnclosure.java:32]: @maxonfjvipon I am not sure that `thread-safe` is correct word for this class. Thread safety means that consistence of instance of this class will not be broken  when different threads will work with this instance. And this is not that we have here.
Here we synchronized static field of the class which means different threads can safely use different instances of the class.

**Comment by levBagryansky** [comment]
@maxonfjvipon Please take a look again. I added concurrency tests and simple tests of `phTracedEnclosure` separately.

**Review by maxonfjvipon — APPROVED** [review]
@levBagryansky LGTM

**Comment by levBagryansky** [comment]
@yegor256 please check

**Pull request closed without merging**